drivers/frequency/adf4382/adf4382.c

```diff
@@ -0,0 +1,1199 @@
+ * @return    		- Result of the writing procedure, error code otherwise.
+ */
+int adf4382_set_phase_pol(struct adf4382_dev *dev, bool polarity)
+{
+	uint8_t pol;
+
+	pol = no_os_field_prep(ADF4382_PHASE_ADJ_POL_MSK, polarity);
+	return adf4382_spi_update_bits(dev, 0x32, ADF4382_PHASE_ADJ_POL_MSK, pol);
+}
+
+/**
+	uint8_t pol;
+ * @brief Gets the polarity of the phase adjust.
+ * @param dev 		- The device structure.
+ * @param polarity 	- The read polarity of the phase.
+ * @return    		- Result of the tesint procedure, negative error code
+ * 			  otherwise.
+ */
+int adf4382_get_phase_pol(struct adf4382_dev *dev, bool *polarity)
+{
+	uint8_t pol;
+
+	pol = adf4382_spi_test_bits(dev, 0x32, ADF4382_PHASE_ADJ_POL_MSK);
+	if (pol < 0)
+		return pol;
+
+	*polarity = pol;
+	return 0;
+}
+
+/**
+ * @brief ADF4382 SPI Scratchpad check.
+ * @param dev 	- The device structure.
+ * @return 	- 0 in case of success or negative error code.
+ */
+static int adf4382_check_scratchpad(struct adf4382_dev *dev)
+{
+	uint8_t scratchpad;
+	int ret;
+
+	ret = adf4382_spi_write(dev, 0x00A, ADF4382_SPI_SCRATCHPAD_TEST);
+	if (ret)
+		return ret;
+
+	ret = adf4382_spi_read(dev, 0x00A, &scratchpad);
+	if (ret)
+		return ret;
+
+	if (scratchpad != ADF4382_SPI_SCRATCHPAD_TEST)
+		return -EINVAL;
+
+	return 0;
+}
+
+/**
+ * @brief Initializes the ADF4382.
+ * @param dev	     - The device structure.
+ * @param init_param - The structure containing the device initial parameters.
+ * @return 	     - 0 in case of success or negative error code.
+ */
+int adf4382_init(struct adf4382_dev **dev,
+		 struct adf4382_init_param *init_param)
+{
+	struct adf4382_dev *device;
+	bool en = true;
+	uint8_t i;
+	int ret;
+
+	device = (struct adf4382_dev *)no_os_calloc(1, sizeof(*device));
+	if (!device)
+		return -ENOMEM;
+
+	ret = no_os_spi_init(&device->spi_desc, init_param->spi_init);
+	if (ret)
+		goto error_dev;
+
+	device->spi_3wire_en = init_param->spi_3wire_en;
+	device->cmos_3v3 = init_param->cmos_3v3;
+	device->ref_freq_hz = init_param->ref_freq_hz;
+	device->freq = init_param->freq;
+	device->ref_doubler_en = init_param->ref_doubler_en;
+	device->ref_div = init_param->ref_div;
+	device->cp_i = init_param->cp_i;
+	device->bleed_word = init_param->bleed_word;
+	device->ld_count = init_param->ld_count;
+	device->phase_adj = 0;
+
+	device->freq_max = ADF4382_RFOUT_MAX;
+	device->freq_min = ADF4382_RFOUT_MIN;
+	device->vco_max = ADF4382_VCO_FREQ_MAX;
+	device->vco_min = ADF4382_VCO_FREQ_MIN;
+	device->clkout_div_reg_val_max = ADF4382_CLKOUT_DIV_REG_VAL_MAX;
+	if (init_param->adf4382a) {
+		device->freq_max = ADF4382A_RFOUT_MAX;
+		device->freq_min = ADF4382A_RFOUT_MIN;
+		device->vco_max = ADF4382A_VCO_FREQ_MAX;
+		device->vco_min = ADF4382A_VCO_FREQ_MIN;
+		device->clkout_div_reg_val_max = ADF4382A_CLKOUT_DIV_REG_VAL_MAX;
+	}
+
+	ret = adf4382_spi_write(device, 0x00, ADF4382_RESET_CMD);
+	if (ret)
+		goto error_spi;
+
+	no_os_udelay(500000);
```

Comment:
is this value datasheet related? if so, please add appropriate comment. and mdelay can be used instead.